### PR TITLE
UCM/RELOC: Optimize relocation patching

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by yosefe** [comment]
OOM on ppc VM
bot:pipe:retest

**Comment by yosefe** [comment]
@hoopoepg @brminich @dmitrygx can you pls take a look?

File: src/ucm/util/reloc.c [hunk 1]
```diff
@@ -96,59 +107,65 @@ static int ucm_reloc_get_aux_phsize()
...
-        found = 0;
-        do {
-            nread = read(fd, buffer, sizeof(buffer));
-            if (nread < 0) {
-                ucm_error("failed to read %lu bytes from %s (ret=%ld): %m",
-                          sizeof(buffer), proc_auxv_filename, nread);
-                break;
-            }
+    if (RUNNING_ON_VALGRIND) {
+        /* Work around a bug caused by valgrind's fake /proc/self/auxv -
```
Comment (dmitrygx) [inline src/ucm/util/reloc.c:122]: `workaround`?

File: src/ucm/util/reloc.c [hunk 2]
```diff
@@ -96,59 +107,65 @@ static int ucm_reloc_get_aux_phsize()
...
+        nread = read(fd, buffer, sizeof(buffer));
+        if (nread < 0) {
+            ucm_error("failed to read %lu bytes from %s (ret=%ld): %m",
+                      sizeof(buffer), proc_auxv_filename, nread);
+            break;
+        }
+
+        count = nread / sizeof(buffer[0]);
+        for (auxv = buffer; (auxv < buffer + count) && (auxv->type != AT_NULL);
+                        ++auxv)
```
Comment (dmitrygx) [inline src/ucm/util/reloc.c:143]: strange alignment

File: src/ucm/util/reloc.c [hunk 3]
```diff
@@ -96,59 +107,65 @@ static int ucm_reloc_get_aux_phsize()
...
+    do {
+        nread = read(fd, buffer, sizeof(buffer));
+        if (nread < 0) {
+            ucm_error("failed to read %lu bytes from %s (ret=%ld): %m",
+                      sizeof(buffer), proc_auxv_filename, nread);
+            break;
+        }
+
+        count = nread / sizeof(buffer[0]);
+        for (auxv = buffer; (auxv < buffer + count) && (auxv->type != AT_NULL);
```
Comment (dmitrygx) [inline src/ucm/util/reloc.c:142]: wrap by `()`
```
... (auxv < (buffer + count)) ...
```

File: src/ucm/util/reloc.c [hunk 4]
```diff
@@ -96,59 +107,65 @@ static int ucm_reloc_get_aux_phsize()
...
+                    ucm_error("phsize is 0");
                 }
+                break;
             }
-        } while ((count > 0) && (phsize == 0));
-
-        if (!found) {
-            ucm_error("AT_PHENT entry not found in %s", proc_auxv_filename);
         }
+    } while ((count > 0) && (phsize == 0));
```
Comment (dmitrygx) [inline src/ucm/util/reloc.c:155]: q: should we check for `!found` instead of `(phsize == 0)`?

File: src/ucm/util/reloc.c [hunk 5]
```diff
@@ -168,128 +185,256 @@ ElfW(Rela) *ucm_reloc_find_sym(void *table, size_t table_size, const char *symbo
...
+    /* modify the relocation to the new value */
+    *entry = patch->value;
+    ucm_debug("symbol '%s' in %s at [%p] modified from %p to %p",
+              patch->symbol, dl_basename, entry, prev_value, patch->value);
+
+    /* store default entry to prev_value to guarantee valid pointers
+     * throughout life time of the process */
+    if (store_prev) {
+        patch->prev_value = prev_value;
+        ucm_debug("'%s' prev_value is %p'", patch->symbol, prev_value);
```
Comment (dmitrygx) [inline src/ucm/util/reloc.c:234]: why need trailing `'` at the end of the string?

File: src/ucm/util/reloc.c [hunk 6]
```diff
@@ -445,6 +583,8 @@ void *ucm_dlopen(const char *filename, int flag)
      * loaded due to dependencies.
      */
 
+    ucm_trace("dlopen(%s) = %p", filename ,handle);
```
Comment (dmitrygx) [inline src/ucm/util/reloc.c:586]: `filename ,handle` -> `filename, handle`

File: src/ucm/util/reloc.c [hunk 7]
```diff
@@ -456,27 +596,89 @@ void *ucm_dlopen(const char *filename, int flag)
...
+        ucm_reloc_orig_dlopen = (ucm_reloc_dlopen_func_t)
+                                ucm_reloc_get_orig(patch->symbol, patch->value);
+        if (ucm_reloc_orig_dlopen == NULL) {
+            ucm_fatal("ucm_reloc_orig_dlopen is NULL");
+        }
+    }
+
+    /* pointer to original dlclose() */
+    if (ucm_reloc_orig_dlclose == NULL) {
+        patch = &ucm_dlopen_reloc_patches[1];
```
Comment (dmitrygx) [inline src/ucm/util/reloc.c:653]: unaligned by `=`

File: src/ucm/util/reloc.c [hunk 8]
```diff
@@ -456,27 +596,89 @@ void *ucm_dlopen(const char *filename, int flag)
...
+    ucm_trace("dlclose(%p)", handle);
+
+    ret = dlinfo(handle, RTLD_DI_LINKMAP, &lm_entry);
+    if (ret != 0) {
+        ucm_warn("dlinfo(handle=%p) failed during dlclose() hook, symbol"
+                 "table may become unreliable", handle);
+    } else {
+        /*
+         * Cleanup the cached information about the library.
+         * NOTE: They library may not actually be unloaded (if its reference
```
Comment (dmitrygx) [inline src/ucm/util/reloc.c:615]: `The library`

File: src/ucm/util/reloc.c [hunk 9]
```diff
@@ -456,27 +596,89 @@ void *ucm_dlopen(const char *filename, int flag)
...
+    { .symbol = "dlclose", .value  = ucm_dlclose }
+};
+
+static void ucm_reloc_get_orig_dl_funcs()
+{
+    ucm_reloc_patch_t *patch;
+
+    /* pointer to original dlopen() */
+    if (ucm_reloc_orig_dlopen == NULL) {
+        patch = &ucm_dlopen_reloc_patches[0];
```
Comment (hoopoepg) [inline src/ucm/util/reloc.c:643]: unaligned ```=``` with line below

**Comment by yosefe** [comment]
failure of r-vmb-fedora16-x86-64
bot:pipe:retest

**Pull request merged**